### ISSUE TITLE
Validation warnings

### ISSUE THREAD

**Issue by jim-obrien-orig** [samuelcolvin/pydantic#939]
# Feature Request: Validation warnings

Please complete:
* OS: debian slim buster
* Python version 3.7
* Pydantic version `import pydantic; print(pydantic.VERSION)`:0.32.2

For some validations, you might want to provide a warning back to the client, rather than an error. A certain threshold might meet the error threshold, but another threshold would just be a warning.  Or if you were to cap values, you might want to warn the client of your app  or service that the value was capped. 

Since this is part of the validation code, it would make sense that the capability to adding warnings should be part of the validation library and not external.

The code below is not prescriptive in terms of method naming or suggesting how this feature would be consumed. It's just to illustrate the idea.

```py
from pydantic import BaseModel, constr, validator

class ModelWithWarningConstraint(BaseModel):
    num: int
    str_field: constr(max_length=5, min_length=2, regex=r'^[a-zA-Z]+', warn_on=['regex'])

    @validator('num')
    def validate_num(cls, v):
        if v > 5:
            raise ValueError("num must be less than 5")
        if v > 3:
            v = 3
        return v, "Number capped at 3"  # add warning(s) to return?

m = ModelWithWarningConstraint(num=4, str_field="a1b2")
# not sure if dict or List[Tuple[str,List[Any[]]
assert len(m.warnings['num']) == 1
assert m.warnings['num'][0] == 'Number capped at 3'
assert len(m.warnings['str_field']) == 1
...
```

**Comment by samuelcolvin** [comment]
Interesting idea but I don't think it would be hard to retrofit this and I think it's too confusing/unusual to add to pydantic.

Happy to consider reopening if lots of people want this.

**Comment by ceesWesseling** [comment]
I have a similar requirement and agree it's too confusing/unusual to add to pydantic.  I am thinking about modelling a warning as a specific type of ValueError. It's dirty and one raises a ValidationError exception while no error may have occured. Codewise it could look like:
`python
try:
   ...
except ValidationError as v:
   only_warnings = True
   for error in v.errors():
     if error.type == MyWarningType:
        logging.warning(error.msg)
     else:
        only_warnings = False
        logging.error(error.msg)
   if not only_warnings:
      raise ...
`

**Comment by internetofjames** [comment]
+1 to this feature request. I'd like to be able to issue warnings that can be gathered up and reported as a single message, similar to how `ValidationErrors` works, but that don't prevent the model from being constructed. I have a use case in a FastAPI app, where model fields typed as an Enum and I want to accept values outside the enum, but want to be notified that these fields received a new value (for example, via log output).

Here's a simple example of a workaround I've tried that gets pretty close to this behavior (I'll mention the problems with this at the end):
```python
class ValidationWarning(Warning):

    def __init__(self, msg: str, field: ModelField = None):
        self.msg = msg
        self.field = field

    def __str__(self):
        if self.field is not None:
            return f"{self.field.alias}: {self.msg}"
        else:
            return super().__str__()


class UnrestrictedEnum(Enum):
   # allows values not defined in the enum to be accepted

    @classmethod
    def allowable_values(cls) -> List:
        return list(map(lambda c: c.value, cls))

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v, field: ModelField):
        if v not in cls.allowable_values():
            warnings.warn(ValidationWarning(f"Unknown value received: {v}", field=field))
            return cls(v)

        else:
            return cls(v)


class Fruit(str, UnrestrictedEnum):
    apple = "apple"
    banana = "banana"


class Vegetable(str, UnrestrictedEnum):
    carrot = "carrot"
    spinach = "spinach"


class Food(BaseModel):
    fruit: Fruit
    vegetable: Vegetable

    def __init__(**kwargs):
         with warnings.catch_warnings(record=True) as validation_warnings:
            super().__init__(**kwargs)

         if validation_warnings:
            warnings.warn(
                f"{len(validation_warnings)} validation warnings in {self.__class__.__name__}\n"
                + "\n".join(str(warn.message) for warn in validation_warnings),
                category=ValidationWarning,
            )
```

Instantiating this model:
```python
f = Food(fruits="strawberry", vegetable="lettuce")
``` 

will result in the following error message being output to `stderr` (with a custom warning formatter to remove the file and line stuff included by the default formatter):
```
ValidationWarning: 2 validation warnings in Food
fruit: Unknown value received: strawberry
vegetable: Unknown value received: lettuce
```


## Issues with this workaround

- As of python 3.10.2, `warnings.catch_warnings` isn't [thread safe](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings)
- I only get the field that had the validation warning in the message, not the full path for nested fields

**Comment by michaeloliverx** [comment]
+1

**Comment by nikokaoja** [comment]
What is the reason why this issue is closed? Is functionality for capturing warnings implemented in pydantic?

**Comment by jim-obrien-orig** [comment]
@samuelcolvin said "Interesting idea but I don't think it would be hard to retrofit this and I think it's too confusing/unusual to add to pydantic.

Happy to consider reopening if lots of people want this."

**Comment by nikokaoja** [comment]
@jim-obrien-orig thanks for reply! I understand reasoning!

**Comment by ceesWesseling** [comment]
In our use of pydantic we have a use case for this. We use pydantic to parse input configurations of software tools, in a way very similar to the messages of a compiler, we like to return:

- as many warnings we can find
- an error if we can not proceed

**Comment by psalvaggio** [comment]
**TLDR**: This is possible on very simple models in a thread-safe manner, but can't capture hierarchical models at all without some help internally from `pydantic` or `pydantic_core`.

I took a stab at this, but I think I have come to the conclusion that is not possible from a user's perspective and would require support from `pydantic`. I started with the solution from @internetofjames, but given that `warnings` is global state, I felt like I couldn't use that. so I started with a simple thread-local solution:
```py
_WARNING_STORAGE = threading.local()

@contextlib.contextmanager
def warning_scope():
    if not hasattr(_WARNING_STORAGE, "data"):
        _WARNING_STORAGE.data = []

    _WARNING_STORAGE.data.append([])
    try:
        yield _WARNING_STORAGE.data[-1]
    finally:
        _WARNING_STORAGE.data.pop()


def warn(warning: Warning) -> None:
    if not hasattr(_WARNING_STORAGE, "data") or len(_WARNING_STORAGE.data) == 0:
        warnings.warn(warning, stacklevel=2)
        return

    _WARNING_STORAGE.data[-1].append(warning)
```

then, defining a custom `ValidationWarning` class:
```py
class ValidationWarning(Warning):
    def __init__(
        self,
        data: dict[str, ty.Union[str, "ValidationWarning"]],
        *,
        model_name: ty.Optional[str] = None,
    ) -> None:
        self._data = data
        self._model = model_name

    def __str__(self):
        def _str_field(
            name: str, warning: ty.Union[str, "ValidationWarning"], *, indent: bool
        ) -> str:
            if isinstance(warning, str):
                return f"{name}:\n  {warning}"
            else:
                return "{}:\n{}".format(
                    name, "\n".join(f"  {line}" for line in str(warning).split("\n"))
                )

        val = ""
        if self._model is not None:
            val = f"{len(self._data)} validation warnings for {self._model}:\n"
        return val + "\n".join(
            _str_field(k, v, indent=self._model is not None)
            for k, v in self._data.items()
        )

    @classmethod
    def field(cls, message: str, *, field_name: str) -> "ValidationWarning":
        return cls({field_name: message})

    @classmethod
    def model(
        cls,
        *,
        model_name: str,
        warnings: ty.Optional[list["ValidationWarning"]] = None,
        message: ty.Optional[str] = None,
    ) -> "ValidationWarning":
        aggregate = {}
        for warn in warnings or []:
            aggregate.update(warn._data)
        if message is not None:
            aggregate["*"] = message
        return cls(aggregate, model_name=model_name)
```
then, I tried hooking into `BaseModel` in one of two ways. I either defined wrappers around `__init__`/`model_validate`/`model_validate_json` that captured warnings, or I replaced the `__pydantic_validator__` with one that captured warnings:
```py
class SchemaValidatorWithWarnings:
    def __init__(self, base: pydantic_core.SchemaValidator) -> None:
        self._base = base

    def _capture_warnings(self, callback: ty.Callable[[], ty.Any]) -> ty.Any:
        with warning_scope() as warns:
            res = callback()

        if warns:
            warn(ValidationWarning.model(model_name=self._base.title, warnings=warns))

        return res

    def __getattr__(self, name):
        return getattr(self._base, name)

    def validate_assignment(self, *args, **kwargs) -> ty.Any:
        return self._capture_warnings(
            lambda: self._base.validate_assignment(obj, *args, **kwargs),
        )

    def validate_python(self, *args, **kwargs) -> ty.Any:
        return self._capture_warnings(
            lambda: self._base.validate_python(*args, **kwargs)
        )

    def validate_json(self, *args, **kwargs) -> ty.Any:
        return self._capture_warnings(
            lambda: self._base.validate_json(*args, **kwargs),
        )

    def validate_strings(self, *args, **kwargs) -> ty.Any:
        return self._capture_warnings(
            lambda: self._base.validate_strings(*args, **kwargs),
        )

class BaseModelWithWarnings(pydantic.BaseModel):
    @classmethod
    def __pydantic_init_subclass__(cls, *args, **kwargs) -> None:
        super().__pydantic_init_subclass__(*args, **kwargs)
        cls.__pydantic_validator__ = SchemaValidatorWithWarnings(
            cls.__pydantic_validator__
        )
```
Finally, I tried a reasonable complex case (nested models, lists, a custom type):
```py
class Username(str):
    pass

class UsernameAdapter:
    @classmethod
    def __get_pydantic_core_schema__(
        cls, source_type: ty.Any, handler: pydantic.GetCoreSchemaHandler
    ) -> pydantic_core.CoreSchema:
        field_name = handler.field_name

        def validator(val: ty.Any, info: pydantic.ValidationInfo):
            if "@" not in val:
                warn(
                    ValidationWarning.field(
                        f"Username must contain @", field_name=field_name
                    )
                )
                val = str(val) + "@unknown"
            return str(val)

        schema = pydantic_core.core_schema.with_info_plain_validator_function(validator)
        return pydantic_core.core_schema.json_or_python_schema(
            json_schema=schema,
            python_schema=schema,
            serialization=pydantic_core.core_schema.plain_serializer_function_ser_schema(
                lambda x: x
            ),
        )

UsernameField = ty.Annotated[Username, UsernameAdapter]

class B(BaseModelWithWarnings):
    users: list[UsernameField]


class A(BaseModelWithWarnings):
    a: int
    b: B

    @pydantic.field_validator("a")
    @classmethod
    def _validate_a(cls, val: int, info: pydantic.ValidationInfo) -> int:
        if val > 0:
            warn(ValidationWarning.field("Value > 0", field_name=info.field_name))
        return val

    @pydantic.model_validator(mode="after")
    def _validate_model(self) -> "A":
        if self.a == 1:
            warn(
                ValidationWarning.model(
                    model_name="A", message="Model level warning"
                )
            )
        return self
```
Doing this, I get the following:
```py
>>> print(A.model_validate({"a": 1, "b": {"users": ["bob"]}}))
3 validation warnings for A:
a:
  Value > 0
users:
  Username must contain @
*:
  Model level warning
a=1 b=B(users=['bob@unknown'])
```
As we can see, I don't get the proper nesting on `users` and there was never any context in which to get the list index. I feel like this solution was a pretty viable approach, but would require a bit more help from the library. If the current path to the value being validated could be passed in via `ValidationInfo`, I feel like I might have been able to make this work.

**Comment by jarekwg** [comment]
+1
We use Pydantic to validate page context before passing it to the frontend.
It's important for the context to be correct, else the frontend can get a surprise, but at the same time raising a hard error here can bring down the whole website for a trivial mismatch. Testing can only get you so far here when context is coming from lots of places and with lots of variance.

We'd love to downgrade the error to just a warning, which we can raise internally without completely locking users out in the interim.

**Comment by luckydonald** [comment]
For rolling out typing to an existing code base it would be really awesome if we could start with an easy mode:
Have it report warnings if we failed to provide the correct type hints. That way the code can continue to run, but we can already use logging (sentry) to collect where we have issues.
Test coverage is not 100% so it makes sense to collect them in the wild - but for that hard errors are a no-go.

**Comment by aKardasz-arista** [comment]
> Interesting idea but I don't think it would be hard to retrofit this and I think it's too confusing/unusual to add to pydantic.
> 
> Happy to consider reopening if lots of people want this.

+1  for adding this as well this would be useful in many ways and as seen through the comments it seems like there is some general buy in.
@samuelcolvin

**Comment by inducer** [comment]
My use case for the feature would be to complain to users about deprecated parts of a schema, e.g. in a before-validator.

I feel that there are a few nearly-orthogonal parts to a possible solution:

- First, being able to report the "current location" so that the user can be told where the issue is. Pydantic does this for error messages, so hopefully this could be straightforward to expose, for example via an additional field in `ValidationInfo`.
- Second, a standard data model for warnings.
- Third, a mechanism to collect those warnings.

IMO, these can be solved more-or-less independently.

**Comment by ipeterov** [comment]
I was looking for something like this to validate tracking URL query params. They [can be huge potentially](https://stackoverflow.com/questions/812925/what-is-the-maximum-possible-length-of-a-query-string), and I have a `max_length` constraint in my DB (255 chars).

I want to truncate the query params to 255, but _if any truncation actually happened_, I want to save the data to DB + report to Sentry so I can see what's happening.

**Comment by andrea-cassioli-maersk** [comment]
+1 It would be a very helpful feature! 

In my use case I get data that need to be patched and there is a need to notify this happening to the user as a warning, not an error.
